### DISCUSSION TITLE
Removing all keys that are not present in a set

### DISCUSSION THREAD

**Discussion by mischnic** [kriszyp/lmdb-js#239] [Q&A]
We want to implement cache garbage collection in Parcel, currently by deleting all LMDB entries that are not listed in a set (which lists all still needed/reachable cache entries). Just iterating over all the keys for this is pretty slow:
```js
    let used = new Set([ /* ... */]);
    for (let k of store.getKeys()) {
      if (!used.has(k)) {
        store.remove(k);
      }
    }
```
Do you know a way to implement this more efficiently?

**Comment by kriszyp** [comment]
Do you have any estimates of roughly what size of store and how many entries are being kept/removed? If most entries are kept, I don't have any good suggestions, although I would have thought that would be pretty fast (I have seen millions of iterations per second on well cached data). However, if more or even close to as many entries are being removed, I wonder if maybe the `remove` is the real bottleneck? I have seen some situations where too many writes in single transaction (this is all in the same transaction because it is in the same event turn) can cause things to slow down a bit (due to larger free-lists to search). Either way, you could add some occasional event turn waits in there (await new Promise(resolve=>setTimeout(resolve))) and see if that makes any difference.
